Lib/Translation/TranslationDb.cs

```diff
@@ -209,28 +210,70 @@ public void SaveLangDbs(string dir, bool justUsed)
-        jw.WriteStartArray();
-        for (var i = 0; i < UsedIds.Count; i++)
-        {
-            jw.WriteStartArray();
-            var idx = (int)UsedIds[i];
-            jw.WriteStringValue(Id2Key[idx].Message);
-            jw.WriteStringValue(Id2Key[idx].Hint);
-            jw.WriteNumberValue(Id2Key[idx].WithParams ? 1 : 0);
-            foreach (var s in Locations[i])
+        switch (_fsAbstraction)
+        {
+            case NativeFsAbstraction:
             {
-                var pathLen = s.LastIndexOf(':', s.LastIndexOf(':') - 1);
-                var p = PathUtils.Subtract(s[..pathLen], projectRoot); 
-                jw.WriteStringValue(p+s[pathLen..]);
+                using var stream = File.Create(PathUtils.Join(dir, "locations.json"));
+                using var jw = CreateJsonWriter(stream);
+                jw.WriteStartArray();
+                for (var i = 0; i < UsedIds.Count; i++)
+                {
+                    jw.WriteStartArray();
+                    var idx = (int)UsedIds[i];
+                    jw.WriteStringValue(Id2Key[idx].Message);
+                    jw.WriteStringValue(Id2Key[idx].Hint);
+                    jw.WriteNumberValue(Id2Key[idx].WithParams ? 1 : 0);
+                    foreach (var s in Locations[i])
+                    {
+                        var pathLen = s.LastIndexOf(':', s.LastIndexOf(':') - 1);
+                        var p = PathUtils.Subtract(s[..pathLen], projectRoot); 
+                        jw.WriteStringValue(p+s[pathLen..]);
+                    }
+
+                    jw.WriteEndArray();
+                }
+
+                jw.WriteEndArray();
+                jw.Flush();
+                break;
             }
+            case FakeFsAbstraction:
+            {
+                using var stream = new MemoryStream();
+                using var jw = CreateJsonWriter(stream);
+                jw.WriteStartArray();
+                for (var i = 0; i < UsedIds.Count; i++)
+                {
+                    jw.WriteStartArray();
+                    var idx = (int)UsedIds[i];
+                    jw.WriteStringValue(Id2Key[idx].Message);
+                    jw.WriteStringValue(Id2Key[idx].Hint);
+                    jw.WriteNumberValue(Id2Key[idx].WithParams ? 1 : 0);
+                    foreach (var s in Locations[i])
+                    {
+                        var pathLen = s.LastIndexOf(':', s.LastIndexOf(':') - 1);
+                        var p = PathUtils.Subtract(s[..pathLen], projectRoot); 
+                        jw.WriteStringValue(p+s[pathLen..]);
+                    }
 
-            jw.WriteEndArray();
-        }
+                    jw.WriteEndArray();
+                }
 
-        jw.WriteEndArray();
-        jw.Flush();
+                jw.WriteEndArray();
+                jw.Flush();
+                _fsAbstraction.WriteAllUtf8(PathUtils.Join(dir, "locations.json"), Encoding.UTF8.GetString(stream.ToArray()));
```

Comment:
It is in UTF-8 already, no need to double decode it and encoded.
Also it is ok to use this "MemoryStream" implementation for all fsAbstraction implementation => not duplicated code is better eventhough it could be teoreticaly less optimal.

Reply:
Fixed. In InMemoryFs I chose to save files as object type. It is converted to string/byte[] as soon as it is read.